File: lib/domain/article.dart
import 'dart:convert';

const Object _articleFieldUnset = Object();

enum ArticleKind { post, page }

class ArticleMetadata {
  const ArticleMetadata({
    this.tags = const <String>[],
    this.categories = const <String>[],
    this.cover,
    this.kind = ArticleKind.post,
    this.templateId,
    this.slug,
    this.extraFrontMatter = const <String, Object?>{},
  });

  final List<String> tags;
  final List<String> categories;
  final String? cover;
  final ArticleKind kind;
  final String? templateId;
  final String? slug;
  final Map<String, Object?> extraFrontMatter;

  ArticleMetadata copyWith({
    List<String>? tags,
    List<String>? categories,
    Object? cover = _articleFieldUnset,
    ArticleKind? kind,
    Object? templateId = _articleFieldUnset,
    Object? slug = _articleFieldUnset,
    Map<String, Object?>? extraFrontMatter,
  }) =>
      ArticleMetadata(
        tags: _unique(tags ?? this.tags),
        categories: _unique(categories ?? this.categories),
        cover: identical(cover, _articleFieldUnset)
            ? this.cover
            : cover as String?,
        kind: kind ?? this.kind,
        templateId: identical(templateId, _articleFieldUnset)
            ? this.templateId
            : templateId as String?,
        slug: identical(slug, _articleFieldUnset) ? this.slug : slug as String?,
        extraFrontMatter: Map<String, Object?>.unmodifiable(
            extraFrontMatter ?? this.extraFrontMatter),
      );

  Map<String, Object?> toJson() => <String, Object?>{
        'tags': tags,
        'categories': categories,
        'cover': cover,
        'kind': kind.name,
        'templateId': templateId,
        'slug': slug,
        'extraFrontMatter': extraFrontMatter,
      };

  factory ArticleMetadata.fromJson(Map<String, Object?> json) =>
      ArticleMetadata(
        tags: _strings(json['tags']),
        categories: _strings(json['categories']),
        cover: json['cover'] as String?,
        kind: ArticleKind.values.byName(json['kind'] as String? ?? 'post'),
        templateId: json['templateId'] as String?,
        slug: json['slug'] as String?,
        extraFrontMatter: Map<String, Object?>.from(
            (json['extraFrontMatter'] as Map?)?.cast<String, Object?>() ??
                <String, Object?>{}),
      );
}

class Article {
  const Article({
    required this.id,
    required this.title,
    required this.body,
    required this.metadata,
    required this.localRevision,
    required this.createdAt,
    required this.updatedAt,
    this.volume,
    this.scheduleAt,
    this.published = false,
  });

  final String id;
  final String title;
  final String body;
  final ArticleMetadata metadata;
  final int localRevision;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? volume;
  final DateTime? scheduleAt;
  final bool published;

  Article copyWith({
    String? title,
    String? body,
    ArticleMetadata? metadata,
    int? localRevision,
    DateTime? updatedAt,
    Object? volume = _articleFieldUnset,
    Object? scheduleAt = _articleFieldUnset,
    bool? published,
  }) =>
      Article(
        id: id,
        title: title ?? this.title,
        body: body ?? this.body,
        metadata: metadata ?? this.metadata,
        localRevision: localRevision ?? this.localRevision,
        createdAt: createdAt,
        updatedAt: updatedAt ?? this.updatedAt,
        volume: identical(volume, _articleFieldUnset)
            ? this.volume
            : volume as String?,
        scheduleAt: identical(scheduleAt, _articleFieldUnset)
            ? this.scheduleAt
            : scheduleAt as DateTime?,
        published: published ?? this.published,
      );

  Map<String, Object?> toJson() => <String, Object?>{
        'id': id,
        'title': title,
        'body': body,
        'metadata': metadata.toJson(),
        'localRevision': localRevision,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
        'volume': volume,
        'scheduleAt': scheduleAt?.toIso8601String(),
        'published': published,
      };

  factory Article.fromJson(Map<String, Object?> json) {
    final Article article = Article(
      id: json['id']! as String,
      title: json['title'] as String? ?? '',
      body: json['body'] as String? ?? '',
      metadata: ArticleMetadata.fromJson(
          (json['metadata'] as Map).cast<String, Object?>()),
      localRevision: json['localRevision'] as int? ?? 1,
      createdAt: DateTime.parse(json['createdAt']! as String),
      updatedAt: DateTime.parse(json['updatedAt']! as String),
      volume: json['volume'] as String?,
      scheduleAt: (json['scheduleAt'] as String?) == null
          ? null
          : DateTime.parse(json['scheduleAt']! as String),
      published: json['published'] as bool? ?? false,
    );
    validateArticle(article);
    return article;
  }

  String encode() => jsonEncode(toJson());
}

Article newArticle(
    {String? id, String title = '', String body = '', DateTime? now}) {
  final DateTime timestamp = now ?? DateTime.now();
  final Article article = Article(
    id: id ?? timestamp.microsecondsSinceEpoch.toString(),
    title: title,
    body: body,
    metadata: const ArticleMetadata(),
    localRevision: 1,
    createdAt: timestamp,
    updatedAt: timestamp,
  );
  validateArticle(article);
  return article;
}

void validateArticle(Article article) {
  if (!RegExp(r'^[a-zA-Z0-9_-]{1,100}$').hasMatch(article.id)) {
    throw const FormatException('Invalid article id');
  }
  if (article.localRevision < 1)
    throw const FormatException('Invalid article revision');
  if (article.body.contains('\u0000'))
    throw const FormatException('Article body contains NUL');
}

List<String> _unique(Iterable<String> values) => values
    .map((String value) => value.trim())
    .where((String value) => value.isNotEmpty)
    .toSet()
    .toList(growable: false);
List<String> _strings(Object? value) => value is List
    ? value.whereType<String>().toList(growable: false)
    : const <String>[];

File: src/domain/article.ts
import { assertDomain } from './errors.ts';
import { articleId, clone, contentHash, newId } from './values.ts';
import type { ArticleId, Revision } from './values.ts';

export type ArticleKind = 'post' | 'page';

export type ArticleMetadata = {
  tags: string[];
  categories: string[];
  cover?: string;
  kind: ArticleKind;
  templateId?: string;
  slug?: string;
  extraFrontMatter: Record<string, unknown>;
};

export type Article = {
  id: ArticleId;
  title: string;
  body: string;
  metadata: ArticleMetadata;
  localRevision: Revision;
  createdAt: string;
  updatedAt: string;
  volume?: string;
  scheduleAt?: string;
  published?: boolean;
};

export function normalizeList(values: readonly string[]): string[] {
  const result: string[] = [];
  const seen = new Set<string>();
  for (const value of values) {
    const normalized = value.trim();
    if (normalized && !seen.has(normalized)) { seen.add(normalized); result.push(normalized); }
  }
  return result;
}

export function createArticle(input: Partial<Omit<Article, 'id' | 'localRevision' | 'createdAt' | 'updatedAt'>> & { id?: string; now?: Date }): Article {
  const now = (input.now ?? new Date()).toISOString();
  const metadata = input.metadata ?? { tags: [], categories: [], kind: 'post' as const, extraFrontMatter: {} };
  const article: Article = {
    id: articleId(input.id ?? newId()),
    title: (input.title ?? '').trim(),
    body: input.body ?? '',
    metadata: {
      tags: normalizeList(metadata.tags ?? []),
      categories: normalizeList(metadata.categories ?? []),
      cover: metadata.cover?.trim() || undefined,
      kind: metadata.kind ?? 'post',
      templateId: metadata.templateId,
      slug: metadata.slug?.trim() || undefined,
      extraFrontMatter: clone(metadata.extraFrontMatter ?? {}),
    },
    localRevision: 1,
    createdAt: input.createdAt ?? now,
    updatedAt: input.updatedAt ?? now,
    volume: input.volume,
    scheduleAt: input.scheduleAt,
    published: input.published ?? false,
  };
  validateArticle(article);
  return article;
}

export function validateArticle(article: Article): void {
  assertDomain(/^[a-zA-Z0-9_-]{1,100}$/.test(article.id) && article.localRevision >= 1, 'Article identity/revision is invalid');
  assertDomain(article.metadata.kind === 'post' || article.metadata.kind === 'page', 'Article kind is invalid');
  assertDomain(!article.body.includes('\0'), 'Article body contains NUL');
  assertDomain(Number.isFinite(Date.parse(article.createdAt)) && Number.isFinite(Date.parse(article.updatedAt)), 'Article dates are invalid');
}

export function updateArticle(article: Article, patch: Partial<Pick<Article, 'title' | 'body' | 'volume' | 'scheduleAt' | 'published' | 'metadata'>>): Article {
  const next = createArticle({ ...clone(article), ...patch, id: article.id, now: new Date() });
  next.localRevision = article.localRevision + 1;
  next.createdAt = article.createdAt;
  next.updatedAt = new Date().toISOString();
  return next;
}

export function articleHash(article: Article): string {
  return contentHash({ title: article.title, body: article.body, metadata: article.metadata, volume: article.volume, scheduleAt: article.scheduleAt });
}

export type ArticleSummary = Pick<Article, 'id' | 'title' | 'updatedAt' | 'localRevision' | 'published' | 'metadata'>;
